### PR TITLE
update cudatoolkit (11.3/11.4), cuddn and cutensor

### PR DESCRIPTION
**Pull request by dguibert**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 137553` at f37414cd run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>21 packages skipped due to time constraints:</summary>
  <ul>
    <li>colmapWithCuda</li>
    <li>cudatoolkit_11_0 (cudaPackages.cudatoolkit_11_0)</li>
    <li>cudatoolkit_11_1 (cudaPackages.cudatoolkit_11_1)</li>
    <li>cudatoolkit_11_2 (cudaPackages.cudatoolkit_11_2)</li>
    <li>cudatoolkit_11_3 (cudaPackages.cudatoolkit_11_3)</li>
    <li>cudnn_cudatoolkit_11 (cudnn_cudatoolkit_11_4)</li>
    <li>cudnn_cudatoolkit_11_0</li>
    <li>cudnn_cudatoolkit_11_1</li>
    <li>cudnn_cudatoolkit_11_2</li>
    <li>cudnn_cudatoolkit_11_3</li>
    <li>...</li>
  </ul>
</details>
<details>
  <summary>2 packages built successfully:</summary>
  <ul>
    <li>cudatoolkit_11 (cudaPackages.cudatoolkit_11 ,cudatoolkit_11_4)</li>
    <li>nccl_cudatoolkit_11</li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/maintainers-missing.md">maintainers-missing</a>

Package does not have a maintainer. Consider adding yourself?

Near pkgs/development/compilers/cudatoolkit/common.nix:239:3:
```
    |
239 |   meta = with lib; {
    |   ^
```
</li>
  </ul>
</details>

**Comment by r-burns** [comment]
Hi, we wanted to add 11.3 and 11.4 first without changing the default version in https://github.com/NixOS/nixpkgs/pull/136655, in case any issues came up. Now that that's merged, we can fix the merge conflicts and go ahead with the full rollout to cuda/cudnn/cutensor 11.4 here if you're up to it.

**Comment by dguibert** [comment]
> Hi, we wanted to add 11.3 and 11.4 first without changing the default version in #136655, in case any issues came up. Now that that's merged, we can fix the merge conflicts and go ahead with the full rollout to cuda/cudnn/cutensor 11.4 here if you're up to it.

Hi, I've rebased this on top thenew master including #136655. Hope it is what you intent.

File: pkgs/development/libraries/science/math/cutensor/default.nix [hunk 1]
```diff
@@ -5,12 +5,12 @@
 
 rec {
   cutensor_cudatoolkit_10_1 = callPackage ./generic.nix rec {
```
Comment (r-burns) [inline pkgs/development/libraries/science/math/cutensor/default.nix:7]: Cutensor 1.3 dropped support for cudatoolkit 10.1 so this doesn't build. Let's:

* keep the cutensor_cudatoolkit_10_1 definition as-is so it still builds
* put the update to 1.3 in cutensor_cudatoolkit_10_2, and have the subsequent versions override from 10_2

File: pkgs/development/libraries/science/math/cutensor/default.nix [hunk 2]
```diff
@@ -14,8 +14,12 @@ rec {
   };
 
   cutensor_cudatoolkit_10_2 = cutensor_cudatoolkit_10_1.override {
+    version = "1.3.1.3";
     libPath = "lib/10.2";
     cudatoolkit = cudatoolkit_10_2;
+    # 1.3.1 is compatible with CUDA 11.0, 11.1, and 11.2:
+    # ephemeral doc at https://developer.nvidia.com/cutensor/downloads
+    sha256 = "sha256-mNlVnabB2IC3HnYY0mb06RLqQzDxN9ePGVeBy3hkBC8=";
```
Comment (samuela) [inline pkgs/development/libraries/science/math/cutensor/default.nix:22]: Why is this necessary? It doesn't seem to be affected by changing `cudnn_cudatoolkit_11` or `cutensor_cudatoolkit_11`

Comment (dguibert) [reply pkgs/development/libraries/science/math/cutensor/default.nix:22]: I don't understand, can you explain how you change cuddn or cutensor?
On top of this branch building cutensor with cudatoolkit 11 depends on this version of cutensor.
```
$ nix-build -A cutensor_cudatoolkit_11
these 2 derivations will be built:
  /nix/store/b9wiirqsjqzcfjps14iyvlpivrd15mlk-cudatoolkit-11.4.1.drv
  /nix/store/5lhpy515j0cqklpma7kjphl9hilksniw-cudatoolkit-11.4-cutensor-1.3.1.3.drv
[...]
```

Comment (samuela) [reply pkgs/development/libraries/science/math/cutensor/default.nix:22]: But the edit here is for `cutensor_cudatoolkit_10_2` not `cutensor_cudatoolkit_11`, right?

Comment (r-burns) [reply pkgs/development/libraries/science/math/cutensor/default.nix:22]: We would typically bump cutensor globally, but cutensor 1.3 drops support for 10.1 so we had to keep 10.1 as-is and bump only 10.2 onward: https://github.com/NixOS/nixpkgs/pull/137553#discussion_r716916285

Comment (samuela) [reply pkgs/development/libraries/science/math/cutensor/default.nix:22]: Ok, I'm a little confused, but I also haven't poked around with this code much so I trust your judgement.

**Comment by r-burns** [comment]
Sorry this stalled out again, let's try to get this merged soon. This still LGTM so I'll test out some representative packages again and allow another day for comments, and if no one requests changes I'll merge this weekend. That way if anything breaks unexpectedly I'll be around to fix/revert.

**Review by r-burns — APPROVED** [review]
Okay, this should be pretty low-impact since our default `cudatoolkit` is still 10.x. The new 11.x toolkits are functional, and the few packages that opt in to cudatoolkit_11 such as jaxlib and terra seem to be working as intended. Thanks for your patience :)
